### DISCUSSION TITLE
Clickjacking- Frameable in Keycloak for /auth/realms/master/protocol/openid-connect/3p-cookies/step1.html

### DISCUSSION THREAD

**Discussion by LakshmiPhani7680** [keycloak/keycloak#29772] [Q&A] [answered]
We want to add X-Frame-Options and CSP headers for the /auth/realms/master/protocol/openid-connect/3p-cookies/step1.html this path of keycloak how to add them? in master settings we have added already the response headers and they are getting reflected for all other paths except this 3p-cookies one, can you please respond on this?
Thank you

**Comment by jonkoops** [comment, marked as answer]
This endpoint does not have any headers to protect it from being embedded elsewhere for performance reasons. If we were to do this we'd have to enumerate every client in a given realm and then match the allowed web origins for each client against the host embedding the iframe. Some users have thousands of clients in their realms, which would make this a possible avenue for a denial of service attack.

We do not allow users to configure this header for this endpoint, as it is extremely easy to set these headers to a value which will render Keycloak inoperable. If you want to configure a security header for this you'll have to do so outside of Keycloak by putting a proxy in front of this endpoint. I'd recommend against using the `X-Frame-Options` header, and instead use the [`frame-ancestors`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors) content security policy directive.

Work is being done behind the scenes to improve this, but we're still in phase where we want to work on a proposal first. Still, this is an issue that should really be resolved, so I will create an issue specifically for the iframe endpoints.

Also, please do not @ mention team members on GitHub unless they are already an active participant in the conversation, as this creates additional notifications and overhead for them.

**Reply by LakshmiPhani7680** [reply]
Hi Jon,
Thanks for the response, so may i know like is this high threat or is it ok to not have the X-Frame-Options or frame-ancestors for this path /auth/realms/master/protocol/openid-connect/3p-cookies/step1.html or /auth/realms/master/protocol/openid-connect/3p-cookies/step2.html for this path.
Thank you

**Reply by jonkoops** [reply]
It's not ideal from a security perspective, but we mitigate it as much as possible by adding using [a pre-flight request](https://github.com/keycloak/keycloak/blob/80de3a0a71bb262c590f3bd411c4017f3db0a789/services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java#L54-L71) to determine if the embedding party is allowed to access the session. This is a compromise between security and performance, but should prevent actors from exploiting this endpoint.

**Reply by LakshmiPhani7680** [reply]
Hi jonkoops,
Thanks for the response, 
as i can see we have X-Frame-Options and frame-ancestors till /auth/master/* i guess but it's not there for this specific endpoint so i think there is not much sensitive info is there for this endpoint right? anyway like it will get patched soon though, but need clarification to submit the report on this issue to my security team.

**Reply by jonkoops** [reply]
Yeah, it should be totally fine not to have those headers there for now. It will not get patched anytime soon, as it is not a priority at this time for the team, but we welcome contributions if your team is up for it. A proposal for a better out of the box Content Security Policy is being worked on, but there are no definitive timelines at this point.

**Reply by LakshmiPhani7680** [reply]
Thank you so much for the quick response, saved my day jon.

**Reply by jonkoops** [reply]
I've logged a follow up issue for this under #29782.
